third_party/nvfuser/csrc/scheduler/reduction_utils.cpp

```diff
@@ -511,6 +521,143 @@ void multiReductionInliner(
   inlineMost();
 }
 
+void propagateTransformation(
+    TensorView* reference_tv,
+    const std::unordered_set<TensorView*>& boundaryNodesSet) {
+  // set an internal boundary to allow different parallel patterns in the fusion
+  struct InternalBoundarySelector : public MaxInfoSpanningTree::Selector {
+    std::unordered_set<TensorView*> _tvs;
+    virtual bool allowC2P(TensorView* from, TensorView* to) override {
+      return _tvs.count(from) == 0;
+    };
+    virtual bool allowP2C(TensorView* from, TensorView* to) override {
+      return _tvs.count(to) == 0;
+    };
+    virtual bool allowSibling(TensorView* from, TensorView* to) override {
+      return true;
+    }
+    InternalBoundarySelector(const std::unordered_set<TensorView*>& tvs)
+        : _tvs(tvs) {}
+  };
+  InternalBoundarySelector ibSelector(boundaryNodesSet);
+  TransformPropagator propagator(reference_tv);
+  MaxRootDomainInfoSpanningTree(reference_tv, &ibSelector)
+      .traverse(&propagator);
+}
+
+void propagateRFactor(
+    TensorView* reference_tv,
+    std::vector<TensorView*> reduction_tvs) {
+  std::unordered_set<int> non_broadcast_rfactor_axes_ir;
+  int non_broadcast_pos_ir = 0;
+  for (const auto i : c10::irange(reference_tv->nDims())) {
+    if (reference_tv->axis((int)i)->isBroadcast()) {
+      continue;
+    }
+    if (reference_tv->axis((int)i)->isReduction() &&
+        reference_tv->axis((int)i)->isRFactorProduct()) {
+      non_broadcast_rfactor_axes_ir.insert(non_broadcast_pos_ir);
+    }
+    non_broadcast_pos_ir++;
+  }
+  for (int i = 1; i < (int)reduction_tvs.size(); i++) {
+    ir_utils::rfactorHelper(
+        reduction_tvs[i],
+        reduction_scheduler_utils::addBackBroadcasts(
+            reduction_tvs[i], non_broadcast_rfactor_axes_ir));
+  }
+}
+
+void propagateParallelization(
```

Comment:
Same here. Can this be used to refactor `multiReductionInliner`? If not, what's the difference?

Reply:
`multiReductionInliner` can be splitted into `propagateTransformation`, `propagateParallelization` and `inlineMost`. If we want to do this refactor, I can do it in another PR.

Reply:
If these new functions are just partial copies of `multiReductionInliner`, it seems rather trivial to simplify `multiReductionInliner` by just calling them, which should reduce code duplicates and makes the code review much easier. 

If they are not just duplicates, I as a reviewer need to understand what the difference is. 

If we can avoid adding technical debts, please take your time. No need to rush.